Schedule option

oschaaf opened the pull request:
Adds a new schedule option to the proto api., which allows specifying a date/time
at which execution should start.

Should also fix #569 by reducing reliance on `Envoy::SystemTime` to a minimum:
Based on the value schedule option, a monotonic time will be computed
and propagated. This makes us use `Envoy::MonotonicTime` again in places
where sub ms resolution is a must have. (The fix will need confirmation, as I
couldn't get `--config=libc++` to work for other reasons.)

dubious90 commented:
@qqustc Please review and assign back to me once ready.

qqustc commented on include/nighthawk/client/output_collector.h:
```diff
@@ -26,7 +29,8 @@ class OutputCollector {
                          const std::map<std::string, uint64_t>& counters,
-                         const std::chrono::nanoseconds execution_duration) PURE;
+                         const std::chrono::nanoseconds execution_duration,
+                         const absl::optional<Envoy::SystemTime>& first_acquisition_time) PURE;
```
Add comment for first_acquisition_time?

dubious90 replied:
will likely be addressed when the comment is added, but I'm not sure i understand what acquisition means in this context.

qqustc reviewed (dismissed):
Thanks Otto!

qqustc commented:
> @qqustc Please review and assign back to me once ready.

LGTM, assigned back to @dubious90 for review.

dubious90 commented on api/client/options.proto:
```diff
@@ -219,4 +220,7 @@ message CommandLineOptions {
   google.protobuf.StringValue latency_response_header_name = 36;
+  // Provide an execution starting date and time. Optional, any value specified must be in the
+  // future.
+  google.protobuf.Timestamp schedule = 105;
```
I feel like schedule here implies the ability to create recurring requests.  How do you feel about scheduled_start?

oschaaf replied:
done here and all other applicable spots.

dubious90 commented on include/nighthawk/common/factories.h:
```diff
@@ -24,7 +24,7 @@ class SequencerFactory {
                               TerminationPredicatePtr&& termination_predicate,
                               Envoy::Stats::Scope& scope,
-                              const Envoy::SystemTime scheduled_starting_time) const PURE;
+                              const Envoy::MonotonicTime scheduled_starting_time) const PURE;
```
Not actionable, just making sure I'm understanding - this is where we're moving back to MonotonicTime to prevent #569 ?

oschaaf replied:
Yes, that is right.

dubious90 commented on source/client/process_impl.cc:
```diff
@@ -179,7 +181,10 @@ void ProcessImpl::createWorkers(const uint32_t concurrency) {
   // TODO(oschaaf): Arguably, this ought to be the job of a rate limiter with awareness of the
```
I think this TODO has become out of date with our ideal, given this restructure.  Do we still want a global rate limiter to do this work anymore, given what we now know?  If not, can we remove this TODO?

dubious90 commented on source/client/process_impl.cc:
```diff
@@ -179,7 +181,10 @@ void ProcessImpl::createWorkers(const uint32_t concurrency) {
   // track-for-future issue.
-  const auto first_worker_start = time_system_.systemTime() + kMinimalWorkerDelay;
+  const Envoy::MonotonicTime monotonic_now = time_system_.monotonicTime();
+  const std::chrono::nanoseconds offset =
```
2. I might be misunderstanding but this ternary seems off to me.  kMinimalWorkerDelay has the worker offsets baked into it, correct?  So, if we are using a scheduled start time, and we aren't using minimal worker delay at all, then we aren't including any worker offsets from each other.

So now, batching behavior of nighthawk acts differently depending on whether or not we are using a schedule which doesn't seem right.

oschaaf replied:
This computes the start of the *first* worker, we will still be injecting an offset between each worker below. I refactored this for clarity, hopefully it's easier to read now.

dubious90 commented on source/client/process_impl.cc:
```diff
@@ -179,7 +181,10 @@ void ProcessImpl::createWorkers(const uint32_t concurrency) {
   // global status quo, which we do not have right now. This has been noted in the
   // track-for-future issue.
-  const auto first_worker_start = time_system_.systemTime() + kMinimalWorkerDelay;
+  const Envoy::MonotonicTime monotonic_now = time_system_.monotonicTime();
```
Three comments here:
1.  This has gotten complex enough (especially with the long comments above to explain what's going on), that I think it might warrant its own privately scoped function.  CalculateWorkerStartTime_ or something?

dubious90 commented on source/client/process_impl.cc:
```diff
@@ -179,7 +181,10 @@ void ProcessImpl::createWorkers(const uint32_t concurrency) {
-  const auto first_worker_start = time_system_.systemTime() + kMinimalWorkerDelay;
+  const Envoy::MonotonicTime monotonic_now = time_system_.monotonicTime();
+  const std::chrono::nanoseconds offset =
+      schedule.has_value() ? schedule.value() - time_system_.systemTime() : kMinimalWorkerDelay;
```
3. We are heavily explaining the workerdelay behavior here in the comment above.  Can we add a one-line comment for how the schedule behavior fits in?  For instance, if I am wrong about point 2 above, we should say why the scheduled workers don't need an offset delay.

dubious90 commented on include/nighthawk/common/rate_limiter.h:
```diff
@@ -33,6 +33,12 @@ class RateLimiter {
+  /**
+   * @return absl::optional<Envoy::SystemTime> Time of the first acquisition, if any.
+   */
+  virtual absl::optional<Envoy::SystemTime> firstAcquisitionTime() const PURE;
```
I might have just missed it in this pull request (if so, please point me to the right file), but while I'm seeing code that writes this field or passes it around, I'm not finding where it's actually used functionally.  What is its intended purpose?

oschaaf replied:
This gets consumed over at https://github.com/envoyproxy/nighthawk/pull/573/files#diff-e528c3973715873e2ec7fcac6a842e5d9a59d6523a77bf1219d295dd4c7285cdR539 and ends up in the output as `execution_start`.

dubious90 commented on test/process_test.cc:
```diff
@@ -178,6 +180,85 @@ TEST_P(ProcessTest, NoFlushWhenCancelExecutionBeforeLoadTestBegin) {
+    // an opportunity to observe passage of simulated time. We increase simulated
+    // time in three steps, to give it an opportunity to start at the wrong time
+    // in case there is an error in the scheduling logic it implements.
+    // Note that these sleeps may seem excessively long, but sanitizer runs may need that.
```
These comments are very helpful.  Thank you.

oschaaf commented:
@dubious90 Thanks for the feedback, I think I addressed it. Extracted the worker start time & offset calculations into their own methods for clarity. Ready for another round!

dubious90 reviewed (approved):
Looks good.  Thanks for the readability changes!  It's really easy to follow what's happening now.
